{
 "metadata": {
  "name": "",
  "signature": "sha256:3ea79410e0b2b586e38108c9a02bb6c018fa52752b8f54656ee69d27ca31522d"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas as pd\n",
      "import math"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Parsing the VCF File"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def find_interval(vcf_pos, interval_list):\n",
      "    \"\"\"Given a list of a chromosome find the interval where the\n",
      "    vcf position is found.\n",
      "    Argument(s):\n",
      "        vcf_pos: The position where the current vcf entry is found.\n",
      "        interval_list: The intervals for the chromosome bands.\n",
      "    \n",
      "    Return(s):\n",
      "        Tuple with the interval in which the vcf position is found.\n",
      "    \"\"\"\n",
      "    \n",
      "    for lower_bound, upper_bound in interval_list:\n",
      "        if lower_bound <= vcf_pos and vcf_pos <= upper_bound:\n",
      "            return (lower_bound, upper_bound)\n",
      "    return \"Roger we have a problem.\"\n",
      "\n",
      "def create_rs_omim_df(file_n = \"snp_omim\"):\n",
      "    \"\"\"Given the snp_omim file from the dbSNP website a dataframe \n",
      "        is made storing rs and mim ids. \n",
      "        Argument(s):\n",
      "            The current working snp_omim file.\n",
      "        Return(s): \n",
      "            Dataframe - Indices: rs id, Columns: mim id\n",
      "    \"\"\"\n",
      "    \n",
      "    with open(file_n) as so_mapping:\n",
      "        so_dict = dict()\n",
      "        for mapping in so_mapping:\n",
      "            mapping = mapping.strip().split()\n",
      "            rs_id, mim_id = mapping[0:2]\n",
      "            so_dict[rs_id] = mim_id\n",
      "        index = [\"{}\".format(i) for i in range(len(so_dict))]\n",
      "        so_series = pd.Series(so_dict)\n",
      "        so_df = pd.DataFrame(so_series, columns = [\"mim_id\"])\n",
      "        so_df.sort(columns = \"mim_id\", axis = 0, inplace = True)\n",
      "    return so_df\n",
      "\n",
      "\n",
      "def create_vcf_interval_dict(file_n = \"17E.vcf\"):\n",
      "    \"\"\"Opening the vfc file and reading the records for the genes affected.\n",
      "        Return(s):\n",
      "            Dictionary - Key(s): position interval, Value(s): cytogenetic position \n",
      "    \"\"\"\n",
      "\n",
      "    create_cytogenetic_dict()\n",
      "    with open(file_n) as vcf_file:\n",
      "        vcf_cytohits_dict = dict()\n",
      "        vcf_file.next()\n",
      "        initiate = vcf_file.next().strip().split()\n",
      "        cur_chrom = initiate[0]\n",
      "        cur_pos = int(initiate[1])\n",
      "        vcf_cytohits_dict[cur_chrom] = dict()\n",
      "        cur_interval = find_interval(cur_pos, cyto_dict[cur_chrom].keys())\n",
      "        vcf_cytohits_dict[cur_chrom][cur_interval] = 1\n",
      "\n",
      "        for vcf_record in vcf_file:\n",
      "            vcf_record = vcf_record.strip().split()\n",
      "            cur_chrom = vcf_record[0].strip()\n",
      "            cur_pos = int(vcf_record[1])\n",
      "            cur_interval = find_interval(cur_pos, cyto_dict[cur_chrom].keys())\n",
      "\n",
      "            if cur_chrom in vcf_cytohits_dict.keys():\n",
      "                if cur_interval in vcf_cytohits_dict[cur_chrom].keys():\n",
      "                    vcf_cytohits_dict[cur_chrom][cur_interval] += 1\n",
      "            else:\n",
      "                vcf_cytohits_dict[cur_chrom] = dict()\n",
      "                vcf_cytohits_dict[cur_chrom][cur_interval] = 1\n",
      "\n",
      "def create_vcf_rs_list(file_n = \"17E.vcf\"):\n",
      "    \"\"\"Reading the vfc file and to obtain a list of all the rs id's.\"\"\"\n",
      "    with open(file_n) as vcf_file:\n",
      "        vcf_rs_ids = list()\n",
      "        vcf_file.next()\n",
      "\n",
      "        for vcf_record in vcf_file:\n",
      "            vcf_record = vcf_record.strip().split()\n",
      "            if vcf_record[2].startswith(\"rs\"):\n",
      "                vcf_rs_ids.append(vcf_record[2].replace(\"rs\", \"\"))\n",
      "\n",
      "    return vcf_rs_ids"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Parsing the cytoband.txt File"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def create_cytogenetic_dict(file_n = \"cytoBand.txt\"):\n",
      "    \"\"\"Making a dictionary for the cytogenetic locations.\"\"\"\n",
      "\n",
      "    with open(\"cytoBand.txt\") as cyto_file:\n",
      "        #initiating the cyto_dictionary\n",
      "        cyto_dict = dict()\n",
      "        initiate = cyto_file.next().strip().split()\n",
      "        prev_chrom = initiate[0].replace(\"chr\",\"\")\n",
      "        for pos in range(1,3):\n",
      "            initiate[pos] = int(initiate[pos])\n",
      "        prev_pos = tuple(initiate[1:3])\n",
      "        cyto_dict[prev_chrom] = dict()\n",
      "        cyto_dict[prev_chrom][prev_pos] = initiate[3]\n",
      "\n",
      "        #Reading the cytoBand.txt file and filling cyto_dict\n",
      "        for cyto_loc in cyto_file:\n",
      "            cyto_loc = cyto_loc.strip().split()\n",
      "            cyto_loc[0] = cyto_loc[0].replace(\"chr\", \"\")\n",
      "\n",
      "            for pos in range(1,3):\n",
      "                cyto_loc[pos] = int(cyto_loc[pos])\n",
      "            cur_chrom = cyto_loc[0]\n",
      "            cur_pos = tuple(cyto_loc[1:3])\n",
      "\n",
      "            if cur_chrom == prev_chrom:\n",
      "                cyto_dict[cur_chrom][cur_pos] = cyto_loc[3]\n",
      "\n",
      "            else:\n",
      "                cyto_dict[cur_chrom] = dict()\n",
      "                cyto_dict[cur_chrom][cur_pos] = cyto_loc[3]\n",
      "                prev_chrom = cur_chrom"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Parsing the omim.txt File"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def create_omim_dict(file_n = \"omim.txt\"):\n",
      "    \"\"\"Opening the OMIM.txt.Z file and creating a dictionary.\n",
      "        Key: omim number, Value: dictionary of key::field, value:: field information.\"\"\"\n",
      "\n",
      "    field_types  = [\"NO\", \"TI\", \"TX\", \"RF\", \"CS\", \n",
      "                    \"CN\", \"CD\", \"ED\", \"SA\"]\n",
      "    collected_fields = [\"NO\", \"TI\", \"TX\"]\n",
      "\n",
      "    with open(file_n) as omim_db:\n",
      "        omim_dict = dict()\n",
      "        record_dict = dict()\n",
      "        field_count = 0\n",
      "        count = 0\n",
      "        cur_field = \"NO\"\n",
      "        prev_field = \"TI\"\n",
      "        prev_line = \"\"\n",
      "        omim_db.next()\n",
      "        omim_db.next()\n",
      "        cur_record_no = omim_db.next().strip()\n",
      "        prev_record_no = cur_record_no\n",
      "        cur_string = \"\"\n",
      "\n",
      "        for line in omim_db:\n",
      "\n",
      "            line = line.strip()\n",
      "\n",
      "            if \"*RECORD*\" not in line:\n",
      "\n",
      "                if any(\"*FIELD* \" + _type in line for _type in field_types):\n",
      "                    cur_field = line.strip().replace(\"*FIELD* \", \"\")\n",
      "                    if cur_field in collected_fields:\n",
      "                        record_dict[prev_field] = cur_string\n",
      "                        cur_string = \"\"\n",
      "\n",
      "                        if cur_field == \"NO\":\n",
      "                            prev_field = \"NO\"\n",
      "                            prev_line = line.strip()\n",
      "\n",
      "                        elif cur_field == \"TI\":\n",
      "                            prev_field = \"TI\"\n",
      "\n",
      "                        elif cur_field == \"TX\":\n",
      "                            prev_field = \"TX\"\n",
      "\n",
      "                    else:\n",
      "                        cur_field = \"NOT COLLECTING\"\n",
      "\n",
      "                elif cur_field == prev_field:\n",
      "\n",
      "                    if cur_field == \"NOT COLLECTING\":\n",
      "                        continue\n",
      "\n",
      "                    if cur_field == \"NO\":\n",
      "                        cur_record_no = line.strip()\n",
      "\n",
      "                    cur_string = cur_string + \" \" + line\n",
      "\n",
      "\n",
      "            elif \"*RECORD*\" in line:\n",
      "                omim_dict[cur_record_no] = record_dict\n",
      "                record_dict = dict()\n",
      "\n",
      "            prev_line = line\n",
      "\n",
      "            field_count += 1\n",
      "            count += 1\n",
      "\n",
      "    return omim_dict\n",
      "\n",
      "def create_omim_df(omim_dict):\n",
      "    NO_list = list()\n",
      "    TI_list = list()\n",
      "    TX_list = list()\n",
      "    #Filling lists from the omim dictionary to\n",
      "    #make a dataframe. \n",
      "    for mim_id, fields in omim_dict.items():\n",
      "        NO_list.append(mim_id)\n",
      "        TI_list.append(omim_dict[mim_id][\"TI\"])\n",
      "        if \"TX\" in omim_dict[mim_id].keys():\n",
      "            TX_list.append(omim_dict[mim_id][\"TX\"])\n",
      "        else:\n",
      "            TX_list.append(\"NA\")\n",
      "\n",
      "    omim_df = pd.concat([pd.Series(NO_list), pd.Series(TI_list), pd.Series(TX_list)], axis = 1)\n",
      "    omim_df.columns = [\"mim_id\", \"TI\", \"TX\"]\n",
      "    return omim_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def create_description_dict(text):\n",
      "    \"\"\"Fills the information of the omim_txt object. \n",
      "        Argument(s): \n",
      "            text - A string resulting from reading in the omim.txt file.\n",
      "    \"\"\"\n",
      "    descriptions = {}\n",
      "    description = text.split()\n",
      "    word_ls = list()\n",
      "    category = \"\"\n",
      "    prev_category = \"\"\n",
      "    description_fields = [\"DESCRIPTION\", \"CLINICAL FEATURES\", \"INHERITANCE\", \"PATHOGENESIS\",\\\n",
      "                          \"CLONING\", \"GENE FUNCTION\", \"MAPPING\", \"MOLECULAR GENETICS\",\\\n",
      "                          \"GENETICS\", \"POPULATION\", \"NOMENCLATURE\", \"ANIMAL MODEL\", \"MODEL\"]\n",
      "    collecting_fields = [\"DESCRIPTION\", \"CLINICAL FEATURES\", \"GENE FUNCTION\"]\n",
      "\n",
      "    for i, word in enumerate(description):\n",
      "\n",
      "        if i + 1 < len(description):\n",
      "            word_plus = word + \" \" + description[i + 1]\n",
      "\n",
      "        if word in description_fields or word_plus in description_fields:\n",
      "\n",
      "            if word in collecting_fields:\n",
      "                category = word \n",
      "\n",
      "            elif word_plus in collecting_fields:\n",
      "                category = word_plus\n",
      "\n",
      "            if category != prev_category and i != 0:\n",
      "                if prev_category in collecting_fields:\n",
      "                    descriptions[prev_category] = \" \".join(word_ls)\n",
      "                prev_category = category\n",
      "                word_ls = []\n",
      "\n",
      "        else:\n",
      "            word_ls += [word]\n",
      "\n",
      "    if len(word_ls) != 0 and prev_category in collecting_fields:\n",
      "        if prev_category in [\"CLINICAL FEATURES\", \"GENE FUNCTION\"]:\n",
      "            word_ls = word_ls[1:]\n",
      "        descriptions[prev_category] = \" \".join(word_ls)\n",
      "    return descriptions"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Writing The Record File"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def write_record_file(file_name = \"index.html\"):\n",
      "    file_handle =open(file_name, \"wt\")\n",
      "\n",
      "    file_handle.write(\"<ul>\\n\")\n",
      "    for i, rec in enumerate(vcf_omim_format_df.iterrows()):\n",
      "        file_handle.write(\"<li>\\n\")\n",
      "\n",
      "        if i > 4:\n",
      "            break\n",
      "        mim_id = rec[1][\"mim_id\"]\n",
      "        ti = rec[1][\"TI\"]\n",
      "        ti = ti.replace(\"*\" + str(mim_id) + \" \", \"\")\n",
      "        ti = ti.replace(\"#\" + str(mim_id) + \" \", \"\")\n",
      "        tx = rec[1][\"TX\"]\n",
      "        hits = vcf_omim_rs_hits[mim_id]\n",
      "        tx_dict = pp.create_description_dict(tx)\n",
      "        rs_ids = mim_id_grouping.get_group(mim_id)\n",
      "        rs_ids_format = []\n",
      "        genemap_mim_ids = genemap_df[\"mimNum\"].tolist()\n",
      "        curr_record = record(mim_id, rs_ids, ti, tx_dict)\n",
      "        curr_record.write_html(file_handle)\n",
      "        \n",
      "        file_handle.write(\"</li>\\n\")\n",
      "    file_handle.write(\"</ul>\\n\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "class record():\n",
      "    def __init__(self, omim_id, rs_ids, TI, TX):\n",
      "        self.omim_id = omim_id\n",
      "        self.rs_ids = rs_ids\n",
      "        self.TI = TI\n",
      "        self.TX = TX\n",
      "        \n",
      "    def write_html(self, file_handle):\n",
      "        file_handle.write(\"Omim Record \" + self.omim_id + \"\\n\")\n",
      "        file_handle.write(\"Associated rs_ids: \" + \", \".join(self.rs_ids) + \"\\n\")\n",
      "        file_handle.write(\"Title: \" + self.TI + \"\\n\")\n",
      "        \n",
      "        if \"DESCRIPTION\" in self.TX:\n",
      "            file_handle.write(\"Description: \" + self.TX[\"DESCRIPTION\"] + \"\\n\")\n",
      "        if \"CLINICAL FEATURES\" in self.TX:\n",
      "            file_handle.write(\"Clinical Features: \" + self.TX[\"CLINICAL FEATURES\"] + \"\\n\")\n",
      "        if \"GENE FUNCTION\" in self.TX:\n",
      "            file_handle.write(\"Gene Function: \" + self.TX[\"GENE FUNCTION\"] + \"\\n\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}